peps/pep-0747.rst

```diff
@@ -569,19 +535,38 @@ Subtyping
 Whether a ``TypeExpr`` value can be assigned from one variable to another is
 determined by the following rules:
 
+Relationship with type
+''''''''''''''''''''''
+
 ``TypeExpr[]`` is covariant in its argument type, just like ``type[]``:
 
 -  ``TypeExpr[T1]`` is a subtype of ``TypeExpr[T2]`` iff ``T1`` is a
    subtype of ``T2``.
 -  ``type[C1]`` is a subtype of ``TypeExpr[C2]`` iff ``C1`` is a subtype
    of ``C2``.
 
-A plain ``type`` can be assigned to a plain ``TypeExpr`` but not the
-other way around:
+An unparameterized ``type`` can be assigned to an unparameterized ``TypeExpr``
+but not the other way around:
 
 -  ``type[Any]`` is assignable to ``TypeExpr[Any]``. (But not the
    other way around.)
 
+Relationship with UnionType
+'''''''''''''''''''''''''''
+
+``TypeExpr[U]`` is a subtype of ``UnionType`` iff ``U`` is a non-empty union type:
+
+-  ``TypeExpr[X | Y | ...]`` is a subtype of ``UnionType``.
+-  ``TypeExpr[Union[X, Y, ...]]`` is a subtype of ``UnionType``.
+-  ``TypeExpr[Optional[X]]`` is a subtype of ``UnionType``.
```

Comment:
It isn't at runtime (currently)

Reply:
Indeed technically I don't think it is, but I don't think the lack of a runtime subtype relationship is observable. §"Interactions with isinstance() and issubclass()" says:

> The ``TypeExpr`` special form cannot be used as any argument to
``issubclass``:

So I'd expect the following behavior:

```
issubclass(TypeExpr[X | Y], UnionType)
TypeError: issubclass() arg 1 must be a class
```

Are there other ways you can think of that a lack of a *runtime* subtype relationship could be observable?

Reply:
```
from types import UnionType
from typing_extensions import Optional, TypeExpr

def f(x: UnionType):
    assert isinstance(x, UnionType)

def g(x: TypeExpr[Optional[int]]):
    f(x)

g(Optional[int])  # boom
```

I'm not sure there is much reason for TypeExpr to ever be a subtype of UnionType; I don't think it will significantly help users of TypeExpr.

Reply:
Clever example. OK it seems that if TypeExpr is to (sometimes) be a subtype of UnionType then there are ways to observe it runtime.

> I'm not sure there is much reason for TypeExpr to ever be a subtype of UnionType; I don't think it will significantly help users of TypeExpr.

I also don't think it will help a ton, but **I think it's necessary for backward compatibility for existing functions that accept `UnionType`** so long as there's no other way to spell "a TypeExpr that is a non-empty union type". Currently §["Rejected Ideas" > "Support pattern matching on type expressions"] does not provide a spelling that can replace existing usage of `UnionType`.

Do you have an alternative suggestion in mind that both:

1. gives a `TypeExpr[int | str]` result for the value expression `int | str` and
2. continues to allow `isinstance('words', int | str)` to pass a type checker?

---

I believe it should be _possible_ to make `TypeExpr[U]` be conditionally considered a subtype of `UnionType` at runtime (via an `isinstance` check) by overriding `__instancecheck__` on the metaclass of `TypeExpr`.

---

Aside: The current implementation of `int | str` gives a `UnionType`, but neither `Union[int, str]` nor `Optional[int]` give `UnionType`s. If we wanted to more-strictly preserve the existing behavior, I'd be open to narrowing the rules in this section to *only* make `TypeExpr[X | Y | ...]` a subtype of `UnionType`:

> -  ``TypeExpr[X | Y | ...]`` is a subtype of ``UnionType``.
> -  ``TypeExpr[Union[X, Y, ...]]`` is *not* a subtype of ``UnionType``.
> -  ``TypeExpr[Optional[X]]`` is *not* a subtype of ``UnionType``.
> -  ``TypeExpr[Never]`` is *not* a subtype of ``UnionType``.
> -  ``TypeExpr[NoReturn]`` is *not* a subtype of ``UnionType``.

Reply:
I think the backwards compatibility problem is an issue of type checker inference that we don't have to specify exactly. It's also  a very limited problem, mostly applying to `isinstance()` which is necessarily special-cased by type checkers anyway.

Reply:
The following code can be written by users today:

```
def accept_union(u: UnionType):
    pass

accept_union(int | str)
```

Are you saying we shouldn’t worry about breaking this code so long as we avoid breaking code related to `isinstance`?

Reply:
I think the way to avoid breaking that code can be up to type checkers. We don't need to prescribe it exactly; different type checkers can use different approaches, and adapt it to changes in how the runtime works. For example, type checkers could store something in their internal representation of a TypeExpr type to indicate the runtime construct that was used to create it.

Reply:
> type checkers could store something in their internal representation of a TypeExpr type to indicate the runtime construct that was used to create it.

Yes, this description is consistent with the implementation approach I had in mypy: A single bit like `is_uniontype`. 

But the overall specification rule being implemented is still:

> ``TypeExpr[X | Y | ...]`` is a subtype of ``UnionType``

I'll update the diff to include only this rule, and not the extraneous ones for `Union` and `Optional`.

Reply:
With only the remaining rule, the code you mentioned before fails (correctly) at type checking time and at runtime:

```
from types import UnionType
from typing_extensions import Optional, TypeExpr

def f(x: UnionType):
    assert isinstance(x, UnionType)  # AssertionError

def g(x: TypeExpr[Optional[int]]):
    f(x)  # ERROR: TypeExpr[Optional[int]] is not a UnionType

g(Optional[int])
```

And similar code involving `TypeExpr[X | Y | ...]` passes (correctly) both at type checking time and at runtime:

```
from types import UnionType
from typing_extensions import Optional, TypeExpr

def f(x: UnionType):
    assert isinstance(x, UnionType)  # OK (runtime)

def g(x: TypeExpr[int | None]):
    f(x)  # OK (type checking time)

g(int | None)
```